Remove PlanPool

abhi1287 opened the pull request:
Follow up from https://github.com/rapyuta-robotics/alica/pull/247
This PR removes PlanPool, introduces RuntimePlanFactory and re-enables all tests.

abhi1287 committed:
Remove PlanPool

veerajsk commented on alica_engine/src/engine/RuleBook.cpp:
```diff
@@ -435,7 +435,7 @@ PlanChange RuleBook::transitionRule(RunningPlan& r)
     const State* nextState = nullptr;
 
-    if (!r.getBasicPlan() || !r.getBasicPlan()->initExecuted()) {
+    if (!r.getBasicPlan()) {
```
don't we still need this because the transition rule is applied before the allocation rule in RuleBook::visit() & so the out transitions will be evalutated before the running plan is activated

abhi1287 replied:
I don't think so.  there is a check above :
if (r.getActiveState() == nullptr)

and there shouldn't be any active state unless the assignment succeed. Guess the issue you fixed here was primarily because a state was assigned by alica thread but its init wasn't executed by the scheduler thread. Let me know if that wasn't the case.

veerajsk replied:
the transition rule is calling moveState which sets the active state to a valid value so it won't be nullptr I think. Also I thought that alica always evaluated the transitions first in order to avoid unnecessary activation of any plans/behaviours in a state. We even relied on this behaviour especially in pause/resume scenarios

veerajsk replied:
the transition rule is calling moveState which sets the active state to a valid value so it won't be nullptr I think. Also I thought that alica always evaluated the transitions first in order to avoid unnecessary activation of any plans/behaviours in a state. We even relied on this behaviour especially in pause/resume scenarios

veerajsk replied:
my bad, I think I may have mistaken the code here. You are right, it shouldn't be required.

veerajsk commented on alica_engine/src/engine/RunningPlan.cpp:
```diff
@@ -570,7 +563,13 @@ void RunningPlan::activate()
+            }
+            _basicPlan = _ae->getRuntimePlanFactory().create(_activeTriple.abstractPlan->getId(), dynamic_cast<const Plan*>(_activeTriple.abstractPlan));
+        }
+        _basicPlan->start(this);
```
is it possible for the same plan to be reactivated? If so we would be calling init again without call onterminate. I think if basic plan exists it should always be stopped & then started (created if necessary)

abhi1287 replied:
How was that handled earlier? I didn't see any explicit plan/behaviour stops in case of reactivation.

veerajsk replied:
There was no handling in running plan & I'm not sure if successive calls to BasicBehaviour::start() will be made by the engine. However, if they are made, the behaviour till now is that BasicBehaviour ignores such calls, i.e. onInit is not called again. But with this refactor  init will be called again which would mean that we violate the condition that init will never be called without terminating the previous run

veerajsk replied:
I fixed it, see commit msg for commit: f127db9a7ffa9cfba12abb09dc44f7b1b16e9715

Dale-Koenig commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -104,7 +102,7 @@ bool AlicaEngine::init(AlicaCreators& creatorCtx)
     _variableSyncModule->init();
     _auth.init();
-    return everythingWorked;
+    return true;
```
If this function always returns true, might as well make it return void

abhi1287 replied:
Well its initializing lots of sub-components and i am sure in some cases it would make sense to return proper error codes. Maybe in a later pr we should fix that rather than changing api.

Dale-Koenig commented on alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -65,6 +65,7 @@ TEST_F(TestBlackBoard, testJsonBlackboardBehavior)
 TEST_F(TestBlackBoard, testNotInheritBlackboardFlag)
 {
     ae->start();
+    ae->getAlicaClock().sleep(alica::AlicaTime::milliseconds(100));
```
Maybe use STEP_UNTIL

abhi1287 replied:
Engine isn't running in step mode in these tests. Out of scope of this pr.

Dale-Koenig replied:
Would be RUN_UNTIL then

Dale-Koenig commented on alica_tests/src/test/test_inherit_blackboard.cpp:
```diff
@@ -28,6 +28,7 @@ TEST_F(TestInheritBlackboard, testInheritBlackboard)
 TEST_F(TestInheritBlackboard, testInheritBlackboardFlag)
 {
     ae->start();
+    ae->getAlicaClock().sleep(alica::AlicaTime::milliseconds(100));
```
STEP_UNTIL

Dale-Koenig commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -61,39 +52,25 @@ void BasicPlan::doRun(void* msg)
 {
-    if (_engine->getTraceFactory()) {
-        _engine->editScheduler().schedule(std::bind(&BasicPlan::traceAssignmentChange, this, assignedEntryPoint, oldUtility, newUtility, numberOfAgents));
+    if (_engine->getTraceFactory() && getTrace()) {
```
IMO if something is worth tracing it is likely worth logging as well at INFO level.
traceAssignmentChange -> logAssignmentChange that both logs in trace and console

abhi1287 replied:
That's already logged in detail in PartialAssignment::operator<<  .
The jagger-client-cpp also had some flag that it can output things on console automatically in addition to sending them to the collector.

veerajsk committed:
Fix reactivation of plan

- Ignore duplicate calls to start & stop in RunnableObject
- Stop the basicPlan on reactivation & allocate a new BasicPlan object
if the plan has changed

veerajsk committed:
Merge branch 'abhi_mega_cleanup' into abhi_mega_cleanup_2
